src/nss/new_nss_oslogin.c

```diff
@@ -241,3 +241,174 @@ int recvline(struct Buffer *const buffer) {
+                        size_t buflen, int *errnop) {
+  int res;
+  struct Buffer mgr;
+  memset(&mgr, 0, sizeof(struct Buffer));
+
+  *errnop = 0;
+  mgr.bufsize = BUFSIZE;
+  mgr.buf = (char *)malloc(BUFSIZE);
+
+  if (dial(&mgr) != 0) {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  // send the verb GETPWUID with the argument <uid>
+  // TODO: validate incoming length of 'uid' fits in 100 char
+  char str[100];
+  sprintf(str, "GETPWUID %d\n", uid);
+  if (send(mgr.socket, str, strlen(str), 0) == -1) {
+      return NSS_STATUS_NOTFOUND;
+  }
+
+  if ((recvline(&mgr)) < 0) {
+    free(mgr.buf);
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  if (mgr.buf[0] == '\n') {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  res = parsepasswd(mgr.buf,result,buffer,buflen);
+  if (res == 0) {
+    return NSS_STATUS_SUCCESS;
+  }
+  *errnop = res;
+  if (res == ERANGE) {
+    return NSS_STATUS_TRYAGAIN;
+  }
+  return NSS_STATUS_NOTFOUND;
+}
+
+static enum nss_status
+_nss_oslogin_getgrnam_r(const char *name, struct group *result, char *buffer,
+                        size_t buflen, int *errnop) {
+  int res;
+  struct Buffer mgr;
+  memset(&mgr, 0, sizeof(struct Buffer));
+
+  *errnop = 0;
+  mgr.bufsize = BUFSIZE;
+  mgr.buf = (char *)malloc(BUFSIZE);
+
+  if (dial(&mgr) != 0) {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  // send the verb GETPWNAM with the argument <name>
+  // TODO: validate incoming length of 'name' fits in 100 char
+  char str[1000];
+  sprintf(str, "GETGRNAM %s\n", name);
+  if (send(mgr.socket, str, strlen(str), 0) == -1) {
+      return NSS_STATUS_NOTFOUND;
+  }
+
+  if ((recvline(&mgr)) < 0) {
+    free(mgr.buf);
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  if (mgr.buf[0] == '\n') {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  res = parsegroup(mgr.buf,result,buffer,buflen);
+  if (res == 0) {
+    return NSS_STATUS_SUCCESS;
+  }
+  *errnop = res;
+  if (res == ERANGE) {
+    return NSS_STATUS_TRYAGAIN;
+  }
+  return NSS_STATUS_NOTFOUND;
+}
+
+static enum nss_status
+_nss_oslogin_getgrgid_r(gid_t gid, struct group *result, char *buffer,
+                        size_t buflen, int *errnop) {
+  int res;
+  struct Buffer mgr;
+  memset(&mgr, 0, sizeof(struct Buffer));
+
+  *errnop = 0;
+  mgr.bufsize = BUFSIZE;
+  mgr.buf = (char *)malloc(BUFSIZE);
+
+  if (dial(&mgr) != 0) {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  // send the verb GETGRGID with the argument <gid>
+  char str[100];
+  sprintf(str, "GETGRGID %d\n", gid);
+  if (send(mgr.socket, str, strlen(str), 0) == -1) {
+      return NSS_STATUS_NOTFOUND;
+  }
+
+  if ((recvline(&mgr)) < 0) {
+    free(mgr.buf);
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  if (mgr.buf[0] == '\n') {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  res = parsegroup(mgr.buf,result,buffer,buflen);
+  if (res == 0) {
+    return NSS_STATUS_SUCCESS;
+  }
+  *errnop = res;
+  if (res == ERANGE) {
+    return NSS_STATUS_TRYAGAIN;
+  }
+  return NSS_STATUS_NOTFOUND;
```

Comment:
This code is repeated except for the string that is sent. Please create a common helper method that takes in the string for sprintf and performs the rest of the common work for easier refactoring at a later time.

Reply:
i think we should not do this - c doesn't lend itself to overloading functions like that well. better to have the four repetitions.